src/basic-noise-qvm.lisp

```diff
@@ -0,0 +1,256 @@
+  ;; depolarization. However, the damping and dephasing noise is
+  ;; time-dependant, so we need to construct the kraus operators for
+  ;; this noise using the elapsed time of the program.  Evolve the
+  ;; elapsed time of the qvm. 
+  ;; Note: This is not an after method because we do not want to
+  ;; invoke CHANNEL-QVM's TRANSITION method.
+  (let* ((gate   (pull-teeth-to-get-a-gate instr))
+         (params (mapcar (lambda (p) (force-parameter p qvm))
+                         (quil:application-parameters instr)))
+         (qubits (mapcar #'quil:qubit-index (quil:application-arguments instr)))
+         (curr-time (+ (avg-gate-time qvm) (elapsed-time qvm)))
+         (damping-ops (loop :for q in qubits
+                            :for q-t1 := (gethash q (t1-vals qvm))
+                            :when q-t1 :collect (damping-kraus-map q-t1 curr-time)))
+         (dephasing-ops (loop :for q :in qubits
+                              :for q-tphi := (gethash q (tphi-vals qvm))
+                              :for q-t1 := (gethash q (t1-vals qvm))
+                              :for q-t2 := (gethash q (t2-vals qvm))
+                              :when (calculate-dephasing-map q-tphi q-t1 q-t2 curr-time) 
+                                :collect it))
+         (depol-ops (loop :for q :in qubits 
+                          :collect (gethash q (depolarization-ops qvm)))))
+    (apply #'apply-gate gate (amplitudes qvm) (apply #'nat-tuple qubits) params)
+    (apply-all-kraus-maps qvm instr (list damping-ops dephasing-ops depol-ops))
+    (setf (elapsed-time qvm) curr-time)
+    (incf (pc qvm))
+    qvm))
+
+(defun calculate-dephasing-map (qubit-tphi qubit-t1 qubit-t2 elapsed-time)
+  "Build the dephasing map. If TPHI is not null, use that to calculate the dephasing map. In the absence of a TPHI value, use T1 and T2 to calculate TPHI if those values exist. T2 is upper bounded by 2 * T1."
+  (cond 
+    ;; First, try to use T-PHI to calculate the dephasing oeprators
+    (qubit-tphi 
+     (dephasing-kraus-map qubit-tphi elapsed-time))
+    ;; Next, if there are values for QUBIT-T1 and QUBIT-T2,
+    ;; calculate T-PHI
+    ((and qubit-t2 qubit-t1) 
+     (dephasing-kraus-map (tphi qubit-t1 qubit-t2) elapsed-time))
+    ;; If either T1 or T2 is missing, an no T-PHI, return no kraus
+    ;; operators.
+    (t nil)))
+
+(defmethod apply-classical-readout-noise ((qvm basic-noise-qvm) (instr quil:measure))
+  ;; Apply the classical readout noise to the state of an BASIC-NOISE-QVM
+  (%corrupt-qvm-memory-with-povm qvm instr (readout-povms qvm)))
+
+(defun tphi (t1 t2)
+  "Calculate t_phi from T1 and T2. t_phi = (2*t1*t2) / (2*t1 + t2). If either T1 or T2 (or both) are 0, TPHI = 0."
+  (assert (< t2 (* 2.0d0 t1)) 
+          (t1 t2)
+          "T2  must be upper bounded by 2*T1")
+  (check-type t1 valid-noise-value)
+  (check-type t2 valid-noise-value)
+  (if (or (= 0.0d0 t1) (= 0.0d0 t2)) 0.0d0)
+  (/ (* 2.0d0 (* t1 t2))
+     (+ t2 (* 2.0d0 t1))))
```

Comment:
For some reason when I do 
```  (if (or (= 0.0d0 t1) (= 0.0d0 t2))
      0.0d0
      (/ (* 2.0d0 (* t1 t2))
         (+ t2 (* 2.0d0 t1)))))```
I get a compiler warning "deleting unreachable code". So I did this instead. It seems to work -- if the `(or ...)` fails, it returns the division computation.

Reply:
The unreachable code note is because the earlier `check-type` against `valid-noise-value` means that t1 and t2 can never be 0.0d0.

Reply:
The "unreachable code" warning is because the type specifier `valid-noise-value` does not include 0.0d0. So the `check-types` above guarantee that `t1` and `t2` are non-zero.

If zero is invalid, you can just delete the `(if ..)` expression.

Reply:
whoops. should have refreshed the page. what zach said.